### PR TITLE
Dash implementation of control charts

### PR DESCRIPTION
**Pull request by TMiguelT**
In reference to #78 and #79 

## Screenshots
Trends look a bit like this:
image

The points show the values of the selected field for each sample, the line shows the mean of these fields, and the filled in section shows 1 standard deviation above and below the mean.

## To Use
Open the website and click the "Data Trend" plot, then select a variable to plot over time

## To Consider
Please look at my code in this pull request, and give feedback on the pros/cons of using Dash as compared to a Jinja template + jQuery

### CODE REVIEW

**Comment by TMiguelT** [comment]
@ewels I realise you might take a while to get to it, but here's my dash PR just so you know about it (not sure if draft PRs give you a notification).

**Comment by ewels** [comment]
Thanks! 👍

File: setup.py [hunk 1]
```diff
@@ -30,11 +30,12 @@
     "markdown>=2.6.11",
     "numpy==1.14.3",
     "passlib==1.7.1",
-    "plotly==2.0.15",
+    "plotly>3.0.0",
     "pyyaml==3.12",
     "SQLAlchemy>=1.1.5",
     "Werkzeug==0.14.1",
     "WTForms>=2.1",
+    "dash @ https://github.com/TMiguelT/dash/archive/master.zip#sha1=e47990eec8fa3f9069040db86ad6eb9a2ad77ce2",
```
Comment (ewels) [inline setup.py]: I don't think that this is being handled properly when I run `python setup.py develop`:

```
Searching for dash@ https://github.com/TMiguelT/dash/archive/master.zip#sha1=e47990eec8fa3f9069040db86ad6eb9a2ad77ce2
Reading https://pypi.python.org/simple/dash/
Downloading https://files.pythonhosted.org/packages/e1/28/c1669c7173cb4b6df44a39122d0e6667cb7eb4a42ca9fa4c3ebd5b90dced/dash-1.0.0.tar.gz#sha256=a5abcf9df0641d8544f2dcd52829891bf2001f697921452b7084694e49b71128
Best match: dash 1.0.0
Processing dash-1.0.0.tar.gz
```

**Comment by TMiguelT** [comment]
Ah, I notice you're using Python 2. Is there any particular reason for that? I think what's happening is that the `import dash` statement I use to import the plotly dash library, is instead importing my `dash.py` file, since it has the same name. Apparently this (dumb) behaviour was removed in Python 3, which is why I never encountered it. Nonetheless, I'll rename that module so the bug can't happen again.

**Comment by ewels** [comment]
> Ah, I notice you're using Python 2. Is there any particular reason for that?

Currently, MultiQC / MegaQC support Py2k. Most python version bugs are specific to py2k and don't occur on py3k, like this one. So I typically set my environment up to be the most likely to fail 😆 

To be fair, MultiQC is very likely to go py3k only pretty soon, mostly because so many of its dependencies are switching to be py3k-only. I was kinda interested in keeping py2k compatibility for MegaQC back in 2017, I'm less fussed now.

> Nonetheless, I'll rename that module so the bug can't happen again.

Seeing as this is an easy one to resolve, doesn't hurt to keep things working on py2k for now I guess. Thanks!

**Comment by TMiguelT** [comment]
Okay, I fixed that and also another weird bug where CSRF was blocking Dash requests. I believe this is fine from a security perspective because of https://github.com/plotly/dash/issues/141

**Comment by ewels** [comment]
That did the trick, thanks! 👍

**Comment by TMiguelT** [comment]
Great! Sidenote, I'm sure this is already obvious but the reason this is still a draft is because I haven't yet implemented the sample filtering that you have in the current trend plot section. I was waiting for something of a go-ahead on Dash before I did that.

**Comment by ewels** [comment]
Ok, it worked great now - couple of speed thoughts. I appreciate that this is a proof of principle, so maybe I'm getting ahead of myself here, but here goes anyway

1. The code looks great, and I'm fine with the idea of using Dash instead of manually creating plotly graphs 👍 
2. I don't want to lose all of the UI for filtering samples - can we keep this?
3. Minor UI issues with the plot itself, eg. only using one colour means it gets very messy when adding more than one data type
4. Legend, tooltips, etc would be nice to customise - probably the same as doing with the manual plotly graph api approach though
5. Would be nice to keep more of the original UI: plot in a div, ability to save the plot as a favourite
6. Some other features such as user authentication is lost, favicon set to the dash icon, etc

For my own reference, the original implementation of this page is here: [api](https://github.com/ewels/MegaQC/blob/1a0b6259cc25d9b6a73cb1007fdf5868706ab8f6/megaqc/api/utils.py#L1098-L1142) and [front-end](https://github.com/ewels/MegaQC/blob/master/megaqc/templates/public/trends.html).

So - for me the advantage of Dash is that we can ditch all of that JavaScript, plus it's more standardised and Python-ish, and easier to get into and read.

The downside is that we may (?) lose the ability to customise the UI quite as much? Or it could be tricky to do so, requiring more hacking in to Dash? This is the reason I mention the above points, as for me I think this is what it will come down to. For example, how difficult will it be to save a plot here to use in a dashboard..

Sorry, this is a of a brain dump and I need to run now before tidying up my thoughts. Hope this all makes sense!

Phil

**Comment by ewels** [comment]
> Great! Sidenote, I'm sure this is already obvious but the reason this is still a draft is because I haven't yet implemented the sample filtering that you have in the current trend plot section.

Sorry, I think we were writing at the same time.. Yes I figured, but I still picked on that 😅 I figure that the filtering should be pretty doable. I'm curious about the plot saving / persistence though - what are your thoughts? I'm also worried that the currently pretty plotly code for generating HTML will get messy and ugly after adding back the UI customisation. Can we add our own dash HTML components?

**Comment by TMiguelT** [comment]
> I don't want to lose all of the UI for filtering samples - can we keep this?
> Would be nice to keep more of the original UI: plot in a div, ability to save the plot as a favourite

See above. I'll add them back once you're happy enough with Dash for me to put in effort to implement this.

> Minor UI issues with the plot itself, eg. only using one colour means it gets very messy when adding more than one data type

Agreed. I can cycle between colours for each additional series

> Legend, tooltips, etc would be nice to customise - probably the same as doing with the manual plotly graph api approach though

Agreed. This should be easy with Dash. The original trend page doesn't do this, does it?

> Some other features such as user authentication is lost, favicon set to the dash icon, etc

I believe this is all doable with Dash (e.g. https://dash.plot.ly/authentication), I just haven't done it yet.

> how difficult will it be to save a plot here to use in a dashboard

I might have to think about this. I'm not sure how the dashboard is currently implemented.

**Comment by TMiguelT** [comment]
> So - for me the advantage of Dash is that we can ditch all of that JavaScript, plus it's more standardised and Python-ish, and easier to get into and read.

Yes, so much less code, much more standardised code (one language), and also the logic becomes viewable in one place, instead of being distributed between the flask view, the flask API, the HTML page, the JS.

**Comment by ewels** [comment]
> > Legend, tooltips, etc would be nice to customise - probably the same as doing with the manual plotly graph api approach though
>
> Agreed. This should be easy with Dash. The original trend page doesn't do this, does it?

Probably not - it was the last page I added and I didn't work on it much. The difficulty in customising plots is my biggest issue with plotly though - for example, HighCharts tooltips can be customised completely with your own HTML, which is much better than anything that I've found for plotly. `</rant>`. But, yes :+1:

> I believe this is all doable with Dash (e.g. https://dash.plot.ly/authentication), I just haven't done it yet.

I don't really want to use the Dash authentication - logging in with plotly is not ideal and the flask auth we already have set up is pretty nice. It would be cool to have google OAuth etc in the future too _(this is what we already use for everything at SciLifeLab - fewer passwords is better)_. I think it should be pretty simple to extend the existing flask auth to this plotly page though - it's just a case of decorating the view function: https://github.com/ewels/MegaQC/blob/1a0b6259cc25d9b6a73cb1007fdf5868706ab8f6/megaqc/public/views.py#L239-L241

[edit: it looks like you're not using the existing flask blueprints, so it will be different. But I'm pretty confident that it shouldn't be that hard to drop in the auth check to what you have here]

> I'm not sure how the dashboard is currently implemented.

The dashboard itself shouldn't need changing, it's more the ability to save a config for a plot and regenerate it at a later date (favourites). This is pretty simple with the existing code, just a case of saving the filtering and fields config. This is all replaced in your dash implementation though, so will need reimplementing.

> What do you mean by UI customization? Do you just mean all of the filtering etc that we had before?

Sort of, though the UI for field selection given by dash is fine. The filtering UI we had previously is more complex though - plus having it hand-written means we can add boxes around stuff, put in buttons where we like (eg. favourite), tweak things here and there. I think this should all be fine, I just have a paranoid view of the future where the Dash code has 15 `dash.Div` functions nested with tonnes of markup for custom HTML + JS. It may be that reusable components + custom react components is the correct way to handle this though 👍 

In summary - I guess what I've been trying to get at all along is that the core plot functionality will be easier to write using Dash, and better. But MegaQC has a deceptively large amount of UI sugar and mini features scattered around. I'm keen not to lose these, as it's these kinds of slick UI elements that tend to make a big difference when using a tool like this on a daily basis. If we can keep them without making the Dash code unwieldy, then great. I'm slightly worried that (a) reimplementing them will take a long time and (b) doing so will lost the benefits of the clean dash code. As we discuss things here, I'm beginning to suspect that your other demo PR of react components is the right way to do it though. Whether it's worth the effort for this almost complete reimplementation of MegaQC is another question 😉 That one is more up to you!

**Comment by TMiguelT** [comment]
> I don't really want to use the Dash authentication

True, I wouldn't want that either, I kind of just grabbed the page as proof that you can do auth. Admittedly their only other option is HTTP auth, but I hope it means it's somewhat extensible to do other kinds of OAuth.

> MegaQC has a deceptively large amount of UI sugar and mini features scattered around. I'm keen not to lose these, as it's these kinds of slick UI elements that tend to make a big difference when using a tool like this on a daily basis

I don't think this is something you really need to worry about. You can write any arbitrary HTML in Plotly, and keep the JS logic in Python. So I should be able to add all those little features without it ruining all the advantages we've been discussing.

Once I've added some of these features back in I think you'll be surprised at how readable it will still be. I'll ping you once I've pushed them.

**Comment by ewels** [comment]
Looking forward to it 👍 💯

**Comment by TMiguelT** [comment]
Okay, I think I admit defeat on Dash. Writing the markup was no issue, and I got it looking awfully like the original:
image.

However, dealing with interaction is a pain:
* You can only create custom components in React, not pure Python, and if we're doing that we might as well use React entirely. My use case for this was making the sample filter box into a re-useable component, because that was a pain point in vanilla JS/HTML. 
* The only way to deal with interactivity is to define a callback in terms of which components it updates, which is a fairly restrictive model of events anyway
* In addition, you can only have 1 callback to update any given component, which restricts your event model even further.

It seemed like you were reconsidering React as a possibility. How are you feeling about it now?

**Pull request closed without merging**